### PR TITLE
Add --log-json option for structured logging output

### LINKED ISSUE #2024: Structured JSON logging
Emit log messages and progress outputs as JSON lines.

This is pretty straightforward in a basic form, but can be enhanced in some ways (eg. provide formatting data as well [1]). We should spec a basic format that we'll adhere to, but leave exact details (eg. messages) open.

A basic implementation can be seen there: https://github.com/enkore/borg/tree/f/logjson

[1]

- Provide formatting data (`log.error("Fu: %s", blar)` -> formatting data = `(blar,) {}`)
- When handling `Error`s, provide the class name (eg. `LockTimeout`, `PathNotAllowed`)
- Provide fixed message-IDs in addition to the more frequently changed log messages themselves (eg. `io-error` when skipping a file due to BackupOSError)

### PR DESCRIPTION
**Pull request by enkore**
fixes #2024

### CODE REVIEW

File: src/borg/logger.py [hunk 1]
```diff
@@ -91,7 +92,11 @@ def setup_logging(stream=None, conf_fname=None, env_var='BORG_LOGGING_CONF', lev
         fmt = '$LOG %(levelname)s %(name)s Remote: %(message)s'
     else:
         fmt = '%(message)s'
-    handler.setFormatter(logging.Formatter(fmt))
+    if json:
+        handler.setFormatter(JsonFormatter(fmt))
+    else:
+        handler.setFormatter(logging.Formatter(fmt))
```
Comment (ThomasWaldmann) [inline src/borg/logger.py]: how about:
```
formatter = JsonFormatter(fmt) if json else logging.Formatter(fmt)
handler.setFormatter(formatter)
```

File: src/borg/logger.py [hunk 2]
```diff
@@ -181,3 +186,31 @@ def critical(self, *args, **kw):
             return self.__logger.critical(*args, **kw)
 
     return LazyLogger(name)
+
+
+class JsonFormatter(logging.Formatter):
+    RECORD_ATTRIBTUES = (
```
Comment (ThomasWaldmann) [inline src/borg/logger.py]: typo

File: src/borg/logger.py [hunk 3]
```diff
@@ -181,3 +186,31 @@ def critical(self, *args, **kw):
...
+        'levelname',
+        'name',
+        'message',
+    )
+
+    # Other attributes that are not very useful but do exist:
+    # processName, process, relativeCreated, stack_info, thread, threadName
+    # msg == message
+    # *args* are the unformatted arguments passed to the logger function, not useful now,
+    # becomes useful if sanitized properly (must be JSON serializable) in the code +
```
Comment (ThomasWaldmann) [inline src/borg/logger.py]: become*

File: src/borg/logger.py [hunk 4]
```diff
@@ -181,3 +186,31 @@ def critical(self, *args, **kw):
...
+    # fixed message IDs are assigned.
+    # exc_info, exc_text are generally uninteresting because the message will have that
+
+    def format(self, record):
+        super().format(record)
+        data = {
+            'type': 'log-message',
+        }
+        for attr in self.RECORD_ATTRIBTUES:
```
Comment (ThomasWaldmann) [inline src/borg/logger.py]: same typo, at least you (your IDE) are consistent. :)

File: src/borg/archive.py [hunk 5]
```diff
@@ -92,19 +94,31 @@ def show_progress(self, item=None, final=False, stream=None, dt=None):
...
-                if space >= 8:
-                    msg += ellipsis_truncate(path, space)
+            if self.output_json:
+                msg = json.dumps({
+                    'type': 'archive-progress',
+                    'osize': self.osize,
+                    'csize': self.csize,
+                    'usize': self.usize,
+                    'nfiles': self.nfiles,
+                    'path': remove_surrogates(item.path if item else ''),
```
Comment (enkore) [inline src/borg/archive.py]: ^- actually should use as_dict + path

**Pull request merged**